### ISSUE TITLE
Validate CNN-compatible hybrid prediction and adaptive-controller readiness

### ISSUE THREAD

**Issue by Sino-Huang** [Sino-Huang/NovPhy#71]
## Category

Implementation readiness and bounded verification. Parent: #1. Follows the completed exploratory #70.

## Why this ticket exists

#70's CNN parser passed its objective audit, but its scoring and training paths requested continuous h15 predictions; it did not directly evaluate adaptive continuous/micro/macro control. The CNN deployment carrier has 18 slots and dimension 236. Older hybrid/controller checkpoints must not be assumed compatible.

This ticket establishes an executable hybrid candidate, not a positive scientific result. Preserve #15/#17/#63/#69/#70 evidence and their interpretations.

## Deadline-scoped work

1. Audit existing mode adapters, shared predictor, symbolic supervision, horizon/mode conditioning, controller, loaders and runtime traces. Report what is trained and reusable versus missing, with estimated training/compute cost before any long run.
2. Complete only the missing integration or training required for the current CNN parser/carrier. Reuse accepted training data and implementations; no encoder sweep, new optimizer, pretrained-backbone search or new final data.
3. Keep one continuous carrier across rollout steps and exactly one selected description mode per decision. Mode/horizon selection must change the executed transition computation, not just a log label or unused output head.
4. Verify that all deployed adapters/readouts/controller parameters are trained, source-bound and loadable. Do not enable random/untrained branches or inject oracle symbols/future outcomes at deployment.
5. Produce the matched fixed-mode control from the SAME pair-capable checkpoint used by the adaptive system. Audit the teacher-forced versus corrected training comparison separately: CNN, data, capacity, mode supervision, seeds and optimizer exposure must match wherever the declared training factor does not require a difference. Existing #70 checkpoints may serve as historical references if these conditions do not hold.

## Acceptance evidence

- Machine-readable readiness report, exact parser/predictor/controller checkpoint identities, carrier/mode/horizon contracts, role provenance, training recipes and compute requirements.
- Shared 236-value carrier and 18-slot vocabulary, with strict incompatible-checkpoint rejection; any required representation change must be proposed before expanding this ticket.
- Regression and bounded integration tests cover trained gradients/heads, meaningful horizon/mode intervention, carrier chaining, controller input restrictions and selected-pair/compute logging.
- A controller that chooses only one pair is reported honestly. Do not force mode variation to pass a test; subsequent advancement requires useful, nondegenerate adaptation.
- Deployment inputs are available agent observations and declared action/interface information only. Engine labels may supervise training or diagnostics, not runtime scoring.
- No new hashes or full-corpus integrity scans. Bounded memory, visible progress/ETA, no-write dry run and bounded real smoke; long training commands are left to the operator.
- Record code revision and ensure required implementation/checkpoint assets are reproducibly identifiable; do not treat an uncommitted local implementation as an archived release.

## Completion and gate

A readiness result does not authorize #64 or claim hybrid superiority. If the existing components cannot provide a valid hybrid candidate within the scoped work, report the missing capabilities/cost and keep the comparison experiment blocked; do not silently substitute continuous-only inference.

## Non-goals

CEM/MPPI/MCTS/Dreamer changes, new perception research, sealed benchmark generation, final access, or overturning previously frozen results.

## Blocked by

- #70 — completed CNN perception and corrected capture infrastructure.

## Next

The matched-grid non-final advancement experiment depends on a validated readiness result from this ticket.

## Unblocks

- #72 — matched-grid non-final comparison, only after validated readiness.

**Comment by Sino-Huang** [comment]
# #71: CNN-compatible hybrid readiness

This is an implementation/training readiness check, not a positive experiment
result. #72 remains blocked until the operator's full run is inspected and its
readiness publication validates. #64/#65 are not authorized.

## Audit and minimal change

The selected #70 corrected checkpoint has a 236-value, 18-slot CNN carrier.
Direct tensor comparison against its seeded initialization confirms that
`micro_adapter`, `macro_adapter`, `micro_head`, and `macro_head` never changed.
Its short-unroll recipe trained continuous h15 only. The #15 hybrid used a
different, 197-value, 15-slot oracle-symbol carrier. Neither is a trained
CNN-compatible adaptive hybrid checkpoint.

The old gameplay wrapper also selects one pair and builds a symbolic request
from the initial observation before its rollout loop. Reusing that request
does not refresh symbols as the predicted carrier changes. Historical code and
evidence are left intact; the new runtime does not use that wrapper.

Reuse the frozen #70 CNN, its carrier codec, accepted #62 training RGB/labels,
and the existing `PairConditioner` / `FiLMBlock` shared-predictor implementation.
No new game collection, encoder training, CEM change, or final data access.

Necessary new components:

- A fixed-slot ordered soft-predicate interface for contact/directed support
  and the two macro state predicates. Small learned carrier readouts replace
  unavailable legacy weights. Availability is supervised with masks; unavailable
  labels are not negatives. Contacts are symmetric; support remains directed.
- Selected-mode inputs are decoded from the current continuous carrier on
  **every** imagined transition. Symbolic tensors are temporary conditioning
  inputs, never a second persistent rollout state. Engine labels enter loss and
  training diagnostics only. No untrained event/duration head is instantiated;
  this is the declared two-predicate macro-state interface, not evidence of
  predicting macro events or validating skipped paths.
- A small joint controller receives only current carrier, interface action,
  and remaining prediction time (lookahead feature capped at 60 steps). Its
  selection executes exactly one of nine `(1,5,15) × (continuous,micro,macro)`
  paths. It cannot request a step beyond the common endpoint. The fixed control
  uses the **same corrected checkpoint**, continuous h15, with no controller call.

This soft ordered interface is not the old capacity-matched #15/#58 graph
adapter or its reliability gate. It is a new, explicitly identified lightweight
CNN-compatible candidate. It cannot rescue or reinterpret the older hypothesis
tests. Neither joint nonseparability nor useful adaptation is established here.

## Frozen, matched training recipe

Use the existing 3,000 training-role lineages only. One-based inventory indices
divisible by five (600 lineages) are excluded from predictor fitting and used
for controller supervision. Both predictors use the other 2,400. The CNN already
used all 3,000; these are not independent perception/generalization estimates.

Each shot supplies up to four uniformly positioned, contiguous 60-fixed-step
windows, including its beginning and end. Parse the preceding RGB frame to
construct motion exactly as at deployment. Windows never cross shots. Short
terminal windows carry their actual length; incomplete requested-horizon targets
are masked, not silently relabeled. Shards preserve lineage, shot, and offset.

Two predictors start from the same seed (20260908), with identical capacity,
data, masks, sampling, and optimizer exposure: 9,000 updates, batch 64, AdamW
1e-4, weight decay 1e-4, clip 1. A group of three fitting lineages visits all nine
pairs before advancing. Every fitting lineage visits every pair; each pair has
1,000 updates. Both arms see the same four local transitions and symbolic
supervision. The corrected arm adds recursive four-step MSE and the declared
carrier-bound penalty; the teacher-forced arm does not. Their compute differs
and is disclosed. This is not a comparison to the unmatched old #70 models.

Controller labels use the frozen corrected model on its 600 excluded fitting
lineages. Duration-weighted dynamic programming minimizes endpoint carrier MSE
times duration, plus 0.0001 times executed linear MACs relative to continuous
h15, plus continuation value. Exact ties use the frozen pair ordering. Only
complete trained horizons are eligible; no artificial mode-balancing labels.

Train 1,800 controller updates, generate one bounded closed-loop aggregation
round, then train another 1,800 on pooled initial/aggregation labels. Aggregation
replaces contexts at actually visited predicted states, retains observed
contexts at other positions, and recomputes DP against unchanged aligned true
targets. Those targets never enter the deployed controller or predictor API.
Use the first window of each controller-training shot, selected without outcomes.

The 24-lineage readiness diagnostics are training-role diagnostics only. Report
pair usage, single-mode behavior, masked symbolic confusion counts, finite
225-step execution and same-endpoint recursive errors. No quality threshold is
tuned here. #72 must separately test useful adaptation and scientific advancement.

## Execution, memory, and evidence

```bash
python -u -m scripts.run_issue_71_hybrid_readiness --dry-run

python -u -m scripts.run_issue_71_hybrid_readiness \
  --smoke-test --device cuda

python -u -m scripts.run_issue_71_hybrid_readiness \
  --run --device cuda \
  2>&1 | tee -a data/issue-71-readiness.log

python -u -m scripts.run_issue_71_hybrid_readiness \
  --validate --device cuda \
  2>&1 | tee -a data/issue-71-validate.log
```

Initialize the `novphy` environment and source `env.sh` first, as usual.
No display or Unity process is needed. `--run` prepares data, trains both
predictors, trains/aggregates the controller, and publishes readiness. Individual
`--prepare`, `--train`, `--train-controller`, and `--publish` stages are available.

Preparation logs each lineage/shot and ETA; training logs every 90 updates and
checkpoints optimizer/sampler state. Completed lineages/models/label shards and
controller rounds are reused. An interrupted controller fitting round restarts
that small round; completed rounds are retained. Rerun the same command to resume.

Only one source capture JSON and a 32-image parser batch are loaded during
preparation. Training holds three lineage shards and one minibatch, not the
image corpus. Three bounded real-lineage measurements took 1.2–2.6 seconds per
lineage; tensor shards were about 0.55–1.1 MB. A brief batch-64 CUDA benchmark
took about 0.03 seconds/update. Observed peak process RSS was about 1.8 GiB;
peak allocated CUDA memory was about 85 MiB. These small warm-cache measurements
suggest roughly **1–3 hours overall**, dominated by preparation, with substantial
uncertainty from shot count and storage contention. Allow several GB of shard
storage. The foreground ETA is more informative once the full run starts.

Artifacts: `.local-artifacts/issue-71-hybrid-readiness-v1/`:

- `plan.json`: source/training/role contracts and exact relevant implementation
  text, existing Git revision, and dirty-worktree disclosure. No new hashes or
  image-integrity pass. Historical source IDs are retained unchanged.
- `teacher_forced.pt`, `corrected.pt`, `controller.pt`: trained state and bindings.
- `readiness.json`: machine-readable readiness, parameter-gradient/exposure
  checks, checkpoint identities, pair traces, diagnostics and compute accounting.
- `smoke.json`: isolated tiny real-data check, explicitly not production evidence.

The linear-MAC accounting includes selected readouts/adapters/trunk/controller;
it is not a full FLOP or end-to-end perception cost claim. Wall time is measured
for preparation/training. #72 must calibrate complete deployment cost before a
matched-compute claim. No engine-derived physical-violation metric is invented
for predicted carriers.

`--validate` recomputes deterministic readiness evidence from the trained
checkpoints and compares it exactly with publication. It rejects incompatible
dimensions, unfinished training, missing task gradients, changed frozen source,
and nonfinite/ineffective pair execution. Single-mode controllers and poor finite
predictions are reported honestly, not silently balanced or called superiority.

Current implementation is local and uncommitted, including prerequisite #70
changes. The source snapshot makes the local implementation identifiable but is
not an archived Git release. Do not close #71 or start #72 merely because the
smoke passed; inspect the full run and complete the repository handoff first.

## Verification completed before handoff

- 115 focused tests passed: 19 new readiness tests plus predictor, temporal
  carrier, CNN parser and #70 parser-repair regressions.
- Public no-write dry run passed against the real source artifacts.
- Public CUDA smoke passed on a real captured training shot: 61 RGB frames,
  18 optimizer updates covering all nine pairs, task gradients reaching every
  deployed parameter tensor, controller labeling/distillation/aggregation-label
  generation, exclusive pair execution, and exact checkpoint reload.
- Full-window preparation was separately exercised on three real lineages,
  including a two-shot lineage. No production shards or model fits were started.
- Resume publication from a completed optimizer checkpoint and the file-backed
  controller-label/aggregation/training/reload path are regression tested.
- `git diff --check` passed. No full-suite claim; no commit/push performed.

Readiness and comparison results are still pending the operator's long run.

**Comment by Sino-Huang** [comment]
## Verified completion: trained executable readiness, not demonstrated hybrid benefit

Independently verified on 2026-09-09 after the operator completed --run and --validate.

The agent reran:
```bash
python -u -m scripts.run_issue_71_hybrid_readiness --validate --device cuda
```
It exited 0 with "exact readiness validation passed; #64 remains unauthorized".

### Verified artifacts and contracts

- Preparation: 3,000 training-role lineage shards, 21,876 windows; 2,400 predictor-fitting lineages and 600 controller-supervision lineages under the frozen split.
- Both predictor checkpoints completed 9,000 updates / 576,000 optimizer examples, exactly 1,000 updates for each of the nine horizon/mode pairs. Both have 1,837,690 parameters.
- Teacher-forced: issue-71-teacher_forced-seed-20260908; corrected: issue-71-corrected-seed-20260908.
- Controller: issue-71-joint-controller-seed-7101-round-1, bound to that corrected checkpoint. Both 1,800-update rounds completed; 600 initial-label and 600 aggregation-label shards exist.
- Strict checkpoint/source-snapshot validation, trained task-gradient/parameter-change checks, 236-value / 18-slot carrier contract, and all pair interventions passed.
- Fixed continuous-h15 and adaptive inference use the SAME corrected predictor checkpoint. The new teacher-forced/corrected controls match the declared architecture/data/supervision/seed/exposure factors; additional recursive training compute is disclosed.
- Runtime uses only current carrier, interface action and remaining steps. Exactly one selected mode executes per decision and symbolic inputs refresh from the predicted carrier. No runtime oracle/future endpoint is provided.
- Full diagnostics and deterministic trace publication recomputed exactly. No new training, collection, hashes, or final access were performed by this verification.

### Results that must not be hidden

Across the 24 training-role diagnostic lineages, the 225-step traces selected:
- h1 continuous: 464 decisions
- h1 macro: 1,046
- h5 micro: 36
- h5 macro: 22
- h15 continuous: 6
- h15 macro: 234

Thus the controller is not single-pair or single-description-mode. This establishes executed variation, NOT useful adaptation.

At the common endpoint used by the recursive diagnostics:
- mean adaptive carrier MSE: 0.1801685206592083
- mean fixed continuous-h15 carrier MSE: 0.0013422288417738553
- adaptive had higher error in all 24/24 cases.

These are development/training-role diagnostics, not a fresh held-out significance test or gameplay result. They are a clear warning against claiming adaptive superiority. The symbolic confusion counts also cover limited first-window states: available macro labels are all steady-state positives and no structure-unstable positives. Zero errors there do not establish broad symbolic or collapse-event competence.

### Disposition and next step

Close #71 as completed trained executable readiness under its explicit scope. Readiness_passed=True is not a scientific advancement decision.

Next is #72's development/freeze phase: inspect this adaptive-versus-fixed weakness and complete the declared quality/compute/feasibility checks before undertaking expensive fresh non-final evaluation. Do not force mode variation, weaken gates, or launch automatic corrective searches to manufacture a positive result.

final_evaluation_opened=false; issue_64_authorized=false. #64/#65 remain unauthorized. Earlier #15/#17/#63/#69/#70 interpretations are unchanged.

Evidence root: .local-artifacts/issue-71-hybrid-readiness-v1/
- readiness.json
- plan.json (frozen contracts and exact relevant source text)
- data.json
- teacher_forced.pt, corrected.pt, controller.pt
- controller-round-0.pt and controller-round-1.pt

Operator logs: data/issue-71-readiness.log and data/issue-71-validate.log.

Repository handoff disclosure: HEAD is 2bfb531968463f28019aa20023ad1f1b93ff1c90, with the #70/#71 implementation still local/uncommitted. The frozen source snapshot identifies the implementation and validation checks it exactly; archived_release remains false. This closure records verified readiness, not a committed/pushed or archived release. No commit or push was performed. Preserve those local source/checkpoint assets for #72 and the later archive handoff.
